File: src/main/java/dev/slne/discord/listener/interaction/menu/DiscordSelectMenuListener.java
package dev.slne.discord.listener.interaction.menu;

import dev.slne.discord.DiscordBot;
import dev.slne.discord.discord.interaction.select.DiscordSelectMenu;
import dev.slne.discord.discord.interaction.select.DiscordSelectMenuManager;
import dev.slne.discord.spring.annotation.DiscordListener;
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;

import javax.annotation.Nonnull;

/**
 * The type Discord select menu listener.
 */
@DiscordListener
public class DiscordSelectMenuListener extends ListenerAdapter {

	private final DiscordSelectMenuManager discordSelectMenuManager;

	/**
	 * Instantiates a new Discord select menu listener.
	 */
	public DiscordSelectMenuListener() {
		this.discordSelectMenuManager = DiscordBot.getInstance().getSelectMenuManager();
	}

	@Override
	public void onStringSelectInteraction(@Nonnull StringSelectInteractionEvent event) {
		DiscordSelectMenu menu = discordSelectMenuManager.getMenu(event.getComponentId());

		if (menu != null) {
			menu.onSelect(event.getInteraction(), event.getValues().stream().map(menu::getOptionByValue).toList());
		}
//		else {
//			event.reply("Die Interaktion ist abgelaufen, oder konnte nicht gefunden werden!").setEphemeral(true)
//				 .queue();
//		}
	}

}

File: src/main/kotlin/dev/slne/discord/listener/interaction/menu/DiscordSelectMenuListener.kt
package dev.slne.discord.listener.interaction.menu

import dev.minn.jda.ktx.events.listener
import dev.slne.discord.discord.interaction.select.DiscordSelectMenuManager
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent

class DiscordSelectMenuListener(jda: JDA) {

    init {
        jda.listener<StringSelectInteractionEvent> { event ->
            val menu = DiscordSelectMenuManager.getMenu(event.componentId)

            if (menu == null) { // TODO: does not work - needed?
//                event.reply("Die Interaktion ist abgelaufen, oder konnte nicht gefunden werden!")
//                    .setEphemeral(true)
//                    .await()

                return@listener
            }

            event.values.mapNotNull { menu.getOptionByValue(it) }
                .let { menu.onSelect(event.interaction, it) }
        }
    }
}
